text/0000-scriptless-buildpacks.md

```diff
@@ -0,0 +1,203 @@
+# Meta
+[meta]: #meta
+- Name: Scriptless Buildpacks
+- Start Date: 7/24/2019
+- CNB Pull Request: (leave blank)
+- CNB Issue: (leave blank)
+- Supersedes: N/A
+
+# Summary
+[summary]: #summary
+
+This is a proposal for new elements in `buildpack.toml` that support the definition of executable buildpacks that contain neither a `bin/detect` nor a `bin/build` (i.e. the `buildpack.toml` can be the only file, but it's still a valid buildpack).
+
+# Motivation
+[motivation]: #motivation
+
+One characteristic of `Dockerfile` that Buildpacks do not replace will is the ability to create a single, one-off file to quickly define a custom build process for an app or project. A `Dockerfile` may contain just a few lines of code, and a developer can generate a complete image.
+
+Conversely, the only quick way to create a custom build for an app using a buildpack is by creating no fewer than three files: `buildpack.toml`, `bin/build`, and `bin/detect`.
+
+We seek to make reduce the barrier to entry for buildpacks when a developer needs to quickly create a custom, one-off build.
+
+# What it is
+[what-it-is]: #what-it-is
+
+Terminology:
+
+- **scriptless**: a buildpack that does not contain the `bin/detect`, `bin/build`, or any other scripts
+
+The target personas for this RFC are the buildpack author and buildpack user who need to create a custom, one-off build or are just getting started with buildpacks and do not want the overhead of a new repo.
+
+We propose the following new tables in `buildpack.toml`:
+
+- `[buildpack.detect]`: (optional) contains elements that are used to perform buildpack detection.
+- `[buildpack.build]`: (optional)  contains elements that are used to perform buildpack building.
+
+Both of these new tables are mutually exclusive with `[[buildpack.order]]`.
+
+## `[buildpack.detect]`
+
+```toml
+[buildpack.detect]
+run = ["<string>"] # bash commands
+requires = [ "<string>"] # defines required entries in the build plan
+provides = [ "<string>" ] # defines provided entries in the build plan
+```
+
+If `buildpack.detect.run` is not defined, and either `buildpack.detect` or `buildpack.build` are defined, then detection will always pass. Otherwise, the exit code of the last `run` command is used.
+
+## `[buildpack.build]`
+
+```toml
+[buildpack.build]
+run = [ "<string>" ]
+
+[[buildpack.build.layers]]
+id = "<string>"
+cache = "<boolean>"
+launch = "<boolean>"
+build = "<boolean>"
+run = [ "<string>" ]
+
+  [buildpack.build.layers.metadata]
+  key = "<string>" # arbitrary elements that will be put directly into the layer metadata
+
+  [buildpack.build.layers.env]
+  KEY = "<string>" # env vars that will be put into `<layer_dir>/env`
```

Comment:
Any thoughts about this compared to the following?
```toml
name = "<string>"
version = "<string>"
```

Reply:
I assume `name = "<string>"` allows us to have a wider of valid names since it's a string. I do like the ergonomics of `KEY= "<string>"` slightly better, but don't feel too strongly either way.

Reply:
I think after some discussion in the WG meeting, we agreed that whatever this is it should match every other `env` section in our TOML files.